src/dataflow/src/sink/tail.rs

```diff
@@ -66,6 +70,31 @@ pub fn tail<G>(
             // though it is slower because it will produce deterministic results since the
             // cursor will always produce rows in the same order.
             results.sort_by_key(|(time, _)| *time);
+
+            if connector.emit_timestamps && !batches.is_empty() {
```

Comment:
you _might_ need to check that `results` is nonempty too? I am not 100% sure but i think if you tail without snapshot you can get nonempty batches that still dont emit anything into `results` which will cause a panic when you unwrap the last element

Reply:
According to frank from slack:

> Because of timely dataflow constraints, DD cannot ship empty batches

so I think this is ok. Added a comment.

Reply:
right. I'm saying that even if the batch was not empty, you might not put anything in `results` because of as of / snapshot constraints. therefore I think you might still need to check that `results` is not empty

Reply:
e.g. i think if you got a batch covering timestamps [0, 4] but were tailing as of 5, `results` would be empty, which may cause a panic if you tried to emit progress messages. haven't tested this

Reply:
> Because of timely dataflow constraints, DD cannot ship empty batches

If at all possible it would be great to avoid relying on this; if DD could ship empty batches it would, and there are (not near-term) changes to timely that would allow this.

If the motivation is just to do this so that you can grab the time with confidence, I would consider using the batch's upper frontier as mentioned in a different comment!

Reply:
```
// Due to timely dataflow constraints, differential dataflow cannot ship empty
// batches, so we are guaranteed that if there's atleast one batch results will
// be non-empty.
```

I don't understand why the implication (if theres at least one batch then results will be nonempty) is true and if it is, could one of you explain why?